ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowDispatch.qll

```diff
@@ -332,98 +343,127 @@ private module Cached {
+      // ```rb
+      // module M
+      //   def instance; end  # <- result
+      // end
+      // M.extend(M)
+      // M.instance           # <- call
+      // ```
+      exists(Module m | result = lookupSingletonMethod(m, method) |
         // ```rb
-        // c = C.new
-        // def c.singleton; end # <- result
-        // c.singleton          # <- call
+        // def C.singleton; end # <- result
+        // C.singleton          # <- call
         // ```
-        // or an `extend`ed instance, e.g.
+        moduleFlowsToMethodCallReceiver(call, m, method)
+        or
         // ```rb
-        // c = C.new
-        // module M
-        //   def instance; end  # <- result
+        // class C
+        //   def self.singleton; end # <- result
+        //   self.singleton          # <- call
         // end
-        // c.extend M
-        // c.instance # <- call
         // ```
-        exists(DataFlow::Node receiver |
-          methodCall(call, receiver, method) and
-          receiver = trackSingletonMethodOnInstance(result, method)
-        )
+        selfInModuleFlowsToMethodCallReceiver(call, m, method)
         or
-        // singleton method defined on a module
-        // or an `extend`ed module, e.g.
         // ```rb
-        // module M
-        //   def instance; end  # <- result
+        // class C
+        //   def self.singleton; end # <- result
+        //   def self.other
+        //     self.singleton        # <- call
+        //   end
         // end
-        // M.extend(M)
-        // M.instance           # <- call
         // ```
-        exists(DataFlow::Node sourceNode, Module m |
-          flowsToMethodCall(call, sourceNode, method) and
-          result = lookupSingletonMethod(m, method)
-        |
-          // ```rb
-          // def C.singleton; end # <- result
-          // C.singleton          # <- call
-          // ```
-          sourceNode = trackModuleAccess(m)
-          or
-          // ```rb
-          // class C
-          //   def self.singleton; end # <- result
-          //   self.singleton          # <- call
-          // end
-          // ```
-          selfInModule(sourceNode.(SsaSelfDefinitionNode).getVariable(), m)
-          or
-          // ```rb
-          // class C
-          //   def self.singleton; end # <- result
-          //   def self.other
-          //     self.singleton        # <- call
-          //   end
-          // end
-          // ```
-          selfInMethod(sourceNode.(SsaSelfDefinitionNode).getVariable(), _, m.getSuperClass*())
-        )
+        selfInMethodFlowsToMethodCallReceiver(call, m, method)
       )
-      or
-      exists(Module superClass, string method |
-        superCall(call, superClass, method) and
-        result = lookupMethod(superClass, method)
-      )
-      or
-      result = yieldCall(call)
     )
+    or
+    exists(Module superClass, string method |
+      superCall(call, superClass, method) and
+      result = lookupMethod(superClass, method)
```

Comment:
To make https://github.com/github/codeql/pull/10747 fully obsolete:

```suggestion
      result = lookupMethod(superClass, method, _)
```

Reply:
I would rather not change the semantics in this PR; only performance.